---
title: "filters VCF for ploidy = 1"
output: html_document
# filters VCF for ploidy = 1
---

#!/usr/bin/env Rscript

#==========================================================#

#####  Filters VCF files

###### Authors : Stefano Mona / Romuald Laso-Jadart / Elise Gay 

###### nov 2023

#==========================================================#


#####  Load Libraries 


```{r}
library(ggplot2)
library(reshape2)
library(gridExtra)
library(stringr)
```

#### Load functions to apply filters on genotype and depth

```{r}
# for filters on sequencing depth 
source("R_functions_ploidy1/fonction_filtre_cover.R")
# for filters on genotypes (heterozygous rate ; homozygous ; missing data)
# Each function can be used to filters individuals or positions based on genotype rates
source("R_functions_ploidy1/Filter_Het.R")
source("R_functions_ploidy1/Filter_Hom_ref.R")
source("R_functions_ploidy1/Filter_Na.R")
```

#### get/set dir
```{r}
getwd()
```

####  Input VCF Files + Description #

####  Example :

  '''
  
  #-----------------------------------------------------------------------------------------------#
    
  # VCF on Y chromosome. Variant calling called with ploidy = 1 
  
  # VCF already filtered for lowqual, indel and repeat. Made from GATK filters on cluster account
  
  # 25 male rats samples (data from Romuald Laso-Jadart dataset)
  
  #-----------------------------------------------------------------------------------------------#
  
  '''

```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

VCF=read.vcfR("data/PPC_M_NC_051357.1_DP100_MQ40_GATK_TAG_Flowqual_Noindels_Norepeat_SNP.vcf.gz")
```

#### Read metadata 
```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

metadata=read.table("metadata/metadata.txt", header=TRUE)
metadata
```

#### 1) Sequencing depth rate visualization

###### Get and format DP table for all position and all individuals

###### get the position sequencing depth
```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

DP<- extract.gt(VCF, element='DP', as.numeric = TRUE) 
head(DP)
summary(DP)
```

###### get mean per column (individu)
```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

moyenne_cover_par_indiv = apply(DP,2,mean,na.rm=T)
moyenne_cover_par_Ind_table=as.data.frame(moyenne_cover_par_indiv)
```

###### summarise the data per individual

```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

head(moyenne_cover_par_Ind_table)
summary(moyenne_cover_par_indiv)
moyenne_cover_par_indiv
```

######  get mean per line (position)
```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

moyenne_cover_par_pos= apply(DP,1,mean,na.rm=T)
moyenne_cover_par_pos_table=as.data.frame(moyenne_cover_par_pos)
```

###### summarise the data per position
```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

head(moyenne_cover_par_pos)
summary(moyenne_cover_par_pos)
```

###### Get the position vector in numeric 
```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

head(rownames(DP))
```

```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

position_vector=as.numeric(str_remove(rownames(DP), "NC_051357.1_"))
head(position_vector)

```

###### Add vector of position in the sequencing depth table 

```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

moyenne_cover_par_pos_table_pos=cbind(moyenne_cover_par_pos_table, position_vector)
head(moyenne_cover_par_pos_table_pos)
dim(moyenne_cover_par_pos_table_pos)
```

#### PLOT detailed DP for each ind or position

##### plot DP by positions

```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

ggplot()+
  geom_point(data=moyenne_cover_par_pos_table_pos, 
             aes(y=moyenne_cover_par_pos_table_pos$moyenne_cover_par_pos,
                 x=moyenne_cover_par_pos_table_pos$position_vector),
             color="black", pch=20, size=1) +
  
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  labs(y="mean depth", x = "PB")+
  ggtitle("Average depth sequencing of SNPs")+
  scale_x_continuous(breaks = seq(0, 18e+06, by = 1000000)) +
  # vizualise only range of sequencing depth detph
  scale_y_continuous(breaks = seq(0, 100, by = 10), limits = c(0,100))

```

##### boxplot DP by individuals

  - format table 
  
```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 

DP_melt=melt(DP)

ggplot()+
  geom_boxplot(data=DP_melt, aes(x=DP_melt$Var2, 
                                 y=DP_melt$value,
                                 colour="purple"),
               color="black") +
  theme(axis.title.x=element_blank())+
  labs(y="SNP sequencing depth", x = "samples")+
  ggtitle("depth sequencing distribution of SNP filtered on 13 males on the R2 region in the Y chromosome") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  # vizualise only range of sequencing depth detph
  scale_y_continuous(breaks = seq(0, 260, by = 20), limits = c(0, 260))
```


#### 2) Vizualisation of chosen genotype frequency

##### Plot count and proportion

- Extract genotype

```{r}
geno<-extract.gt(VCF,element="GT",mask=F,as.numeric=F,return.alleles = F,
                   convertNA = F,extract = T)
geno_table=as.data.frame(geno)
head(geno_table)
```

- Raw Count of genotype 

```{r}
# In rowsum and colsum the number of '.' '1' and '0'and compute genotype count 

# By position
sum_pos_geno_table = as.data.frame(rowSums(geno==".")) # put ".", "0" or "1" depending on the genotype you want to count
head(sum_pos_geno_table)
dim(sum_pos_geno_table)
```

```{r}
# By individuals
sum_ind_geno_table = as.data.frame(colSums(geno=="."))
dim(sum_ind_geno_table)
summary(sum_ind_geno_table)
```

- Genotype frequency

```{r}
# By position
Prop_Pos_geno=((sum_pos_geno_table/dim(geno_table)[2])*100)
head(Prop_Pos_geno)
summary(Prop_Pos_geno)
Prop_Pos_geno_table=as.data.frame(Prop_Pos_geno)
```

```{r}
# Get the position vector in numeric 
head(rownames(Prop_Pos_geno_table))
position_vector=as.numeric(str_remove(rownames(Prop_Pos_geno_table), "NC_051357.1_"))
```

```{r}
# Add vector of position in the sequencing depth table 
Prop_Pos_geno_table_pos=cbind(Prop_Pos_geno_table, position_vector)
head(Prop_Pos_geno_table_pos)
dim(Prop_Pos_geno_table_pos)
```

```{r}
# By individuals
Prop_Ind_geno=((sum_ind_geno_table/dim(geno_table)[1])*100)
head(Prop_Ind_geno)
summary(Prop_Ind_geno)
Prop_Ind_geno_table=as.data.frame(Prop_Ind_geno)
Prop_Ind_geno_table

```

#### PLOT genotype frequencies

```{r}
# By position
ggplot()+
  geom_point(data=Prop_Pos_geno_table_pos, 
             aes(y=Prop_Pos_geno_table_pos$`rowSums(geno == ".")` , 
                 x=Prop_Pos_geno_table_pos$position_vector), 
             color="purple",
             size=1)+
  labs(x="pos", y = "Prop '.' site") +
  ggtitle("Prop '.' site by position on the Y chromosome") +
  scale_x_continuous(breaks = seq(0, position_vector[length(position_vector)], 
                                  by = 1000000)) +
  scale_y_continuous(breaks = seq(0, 100, by = 10), limits = c(0,100)) +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1))
```


```{r}
# By indivduals
ggplot()+
  geom_point(data=Prop_Ind_geno_table, 
             aes(y=Prop_Ind_geno_table$`colSums(geno == ".")` , 
                 x=row.names(Prop_Ind_geno_table)),
             color="purple",
             size=2) +
  labs(x="Ind", y = "Prop '.' site") +
  ggtitle("Prop '.' site by individuals on the Y chromosome") + 
  theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust=1)) +
  scale_y_continuous(breaks = seq(0, 100, by = 10), limits = c(0,100))

```

#### 3) Apply filters 

##### Example of filters by individuals 
```{r}
colnames(VCF@gt)
VCF_subset = VCF[,c("FORMAT","ALK20","ANR0085_S11","ANR0086_S14","CAL79","CN001","CN002")]
```

##### Functions description

'''

- For DP filters

	# ARGUMENTS : fun_filtre_cover<-function(vcf,taux_min_cover_pos,taux_max_cover_pos,taux_min_cover_ind,taux_max_cover_ind){

	# taux_min_cover_pos : int, min average DP by positions

	# taux_max_cover_pos : int, max average DP by positions

	# taux_min_cover_ind : int, min average DP by individuals

	# taux_max_cover_ind : int, max average DP by individuals

- For Na filters

	# ARGUMENTS :function(vfc,rate_na_max_POS,rate_na_max_Ind)

	# Method : 

	# sum of "genotype" in vcf@gt has to be inferior or equal to the rate chosen in arg (which(rate_na_ind<=rate_na_max_Ind))

	# The filters is first applied on individuals and then on position



- For rate of heterozygous sites filters

	# ARGUMENTS : 

	# rate_het_max_POS (double) : Max percent of heterozygous site in one position

	# rate_het_max_Ind (double) : Max percent of heterozygous site in one individus

	# rate_het_min_pos (int) : Min nb of heterozygous site in one position

	# Method : 

	# sum of "genotype" in vcf@gt has to be inferior or equal to the rate chosen in arg (which(rate_het_ind<=rate_het_max_Ind))

	# The filters is first applied on individuals and then on position

'''

#### Run the loop

- examples of filters on raw vcf and vcf subset
```{r}
dir_data="data/"
VCF_lists=c("VCF", "VCF_subset")
```

- Run filters on the VCF list

```{r}
for (vcf in VCF_lists){
  name_vcf=get(vcf)
  # filter for DP 
  VCF_DP_10_100<-fun_filtre_cover(name_vcf, 10, 100, 10, 200)
  
  # filter for Na (0% by position)
  VCF_DP_10_100_Na0=Filters_na_Ind_Pos(VCF_DP_10_100,0.0,0.0)
  
  # other filters with NA < 20% by positions and 30% by individuals 
  VCF_DP_10_100_Na20_30=Filters_na_Ind_Pos(VCF_DP_10_100,20.0,30.0)
  
  # save vcf filtered
  file_name_na=paste(dir_data,"Rats_Y_P1_DP100_Flowqual_Noindels_Norepeat_DP10_100_Na0_", vcf, "_vcf.gz", sep = "")
  write.vcf(VCF_DP_10_100_Na0, paste(dir_data,"Rats_Y_DP10_100_Na0.vcf.gz", sep = ""))
  write.vcf(VCF_DP_10_100_Na20_30, paste(dir_data,"Rats_Y_DP10_100_Na20_30.vcf.gz", sep = ""))
}
```

- Check for data 
```{r}
dir(dir_data)
```


#### To check your filters, re-run the script on the filtered vcfs



